File: youtube/youtube.php
<?php
declare(strict_types = 1);
extension_loaded('openssl') or die('openssl');
require_once 'cove/helper.php';

function format_number(float $n): string {
   $n2 = (int)(log10($n) / 3);
   return sprintf('%.3f', $n / 1e3 ** $n2) . ['', ' K', ' M', ' B'][$n2];
}

function youtube_info(string $id_s): object {
   # part 1
   $info_s = 'https://www.youtube.com/get_video_info?video_id=' . $id_s;
   echo $info_s, "\n";
   # part 2
   $get_s = file_get_contents($info_s);
   parse_str($get_s, $get_m);
   # part 3
   $resp_s = $get_m['player_response'];
   # part 4
   return json_decode($resp_s)->microformat->playerMicroformatRenderer;
}

function youtube_views(object $info_o): string {
   $views_n = (int)($info_o->viewCount);
   $old_o = DateTime::createFromFormat('!Y-m-d', $info_o->publishDate);
   $new_o = new DateTime;
   $diff_n = $new_o->diff($old_o)->days / 365;
   $rate_n = $views_n / $diff_n;
   $rate_s = format_number($rate_n);
   if ($rate_n > 8_000_000) {
      return 'RED ' . color_red($rate_s);
   }
   return 'GREEN ' . color_green($rate_s);
}

File: music-database/youtube/youtube.go
package youtube

import (
   "math"
   "net/url"
)

func NumberFormat(n float64) string {
   n2 := int(math.Log10(n)) / 3
   n3 := n / math.Pow10(n2 * 3)
   return fmt.Sprintf("%.3f", n3) + []string{"", " k", " M", " B"}[n2]
}

func GetContents(s string) (string, error) {
   o, e := http.Get(s)
   if e != nil {
      return "", e
   }
   y, e := ioutil.ReadAll(o.Body)
   if e != nil {
      return "", e
   }
   return string(y), nil
}

function Info(id_s string) (Map, error) {
   info_s := "https://www.youtube.com/get_video_info?video_id=" + id_s
   println(info_s)
   query_s, e := GetContents(info_s)
   if e != nil {
      return nil, e
   }
   m, e := url.ParseQuery(query_s)
   if e != nil {
      return nil, e
   }
   resp_s := m["player_response"]
   return json_decode($resp_s)->microformat->playerMicroformatRenderer;
}

function youtube_views(object $info_o): string {
   $views_n = (int)($info_o->viewCount);
   $date_o = DateTime::createFromFormat('!Y-m-d', $info_o->publishDate);
   $sec_n = seconds(new DateTime, $date_o);
   $rate_n = $views_n / ($sec_n / 60 / 60 / 24 / 365);
   $rate_s = format_number($rate_n);
   if ($rate_n > 8_000_000) {
      return 'RED ' . color_red($rate_s);
   }
   return 'GREEN ' . color_green($rate_s);
}
